crypto/tls/tls.go

```diff
@@ -109,29 +122,42 @@ func (cfg *ClientConfig) GetTLSConfig() (*tls.Config, error) {
 		config.RootCAs = caCertPool
 	}
 
-	// Read Client Certificate
-	if cfg.CertPath != "" || cfg.KeyPath != "" {
-		if cfg.CertPath == "" {
-			return nil, errCertMissing
-		}
-		if cfg.KeyPath == "" {
-			return nil, errKeyMissing
+	loadCert := func() (*tls.Certificate, error) {
+		// not used boolean, assumed if this is called is because we already configured TLS Client certificates.
+		_, err := cfg.validateCertificatePaths()
+		if err == nil {
+			cert, err := reader.ReadSecret(cfg.CertPath)
+			if err != nil {
+				return nil, errors.Wrapf(err, "error loading client cert: %s", cfg.CertPath)
+			}
+			key, err := reader.ReadSecret(cfg.KeyPath)
+			if err != nil {
+				return nil, errors.Wrapf(err, "error loading client key: %s", cfg.KeyPath)
+			}
+
+			clientCert, err := tls.X509KeyPair(cert, key)
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to load TLS certificate %s,%s", cfg.CertPath, cfg.KeyPath)
+			}
+			return &clientCert, nil
 		}
+		return nil, err
+	}
 
-		cert, err := reader.ReadSecret(cfg.CertPath)
-		if err != nil {
-			return nil, errors.Wrapf(err, "error loading client cert: %s", cfg.CertPath)
-		}
-		key, err := reader.ReadSecret(cfg.KeyPath)
-		if err != nil {
-			return nil, errors.Wrapf(err, "error loading client key: %s", cfg.KeyPath)
+	useClientCerts, err := cfg.validateCertificatePaths()
+	if err != nil {
+		return nil, err
+	}
+
+	if useClientCerts {
```

Comment:
It makes more sense to me to only validate the configuration parameters before setting `config.GetClientCertificate`, unless you want to fully handle configuration parameters potentially changing:

```suggestion
	// Read Client Certificate
	if cfg.CertPath != "" || cfg.KeyPath != "" {
		if cfg.CertPath == "" {
			return nil, errCertMissing
		}
		if cfg.KeyPath == "" {
			return nil, errKeyMissing
		}
```